File: Alfred/CS2040C Data Structures and Algorithms/CS2040C Data Structures and Algorithms/Lectures/Lecture12/shortestpath1_dijkstra_semilive.java
// I put one character typo in this live code, so submitting it verbatim will give you WA

import java.io.*;
import java.util.*;

class shortestpath1_dijkstra_semilive {
  private static final int INF = 1000000000; // good idea to define INF as a constant, so that you don't encounter silly bug involving accidental zero(es)

  public static void main(String[] args) throws Exception {
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    PrintWriter pw = new PrintWriter(System.out);

    while (true) {
      String[] token = br.readLine().split(" ");
      int n = Integer.parseInt(token[0]), m = Integer.parseInt(token[1]), q = Integer.parseInt(token[2]), s = Integer.parseInt(token[3]);
      if (n == 0 && m == 0 && q == 0 && s == 0) break;
      Vector<Vector<IntegerPair>> AL = new Vector<Vector<IntegerPair>>();
      for (int u = 0; u < n; u++) {
        Vector<IntegerPair> neighbors = new Vector<IntegerPair>();
        AL.add(neighbors);
      }
      while (m-- > 0) {
        token = br.readLine().split(" ");
        int u = Integer.parseInt(token[0]), v = Integer.parseInt(token[1]), w = Integer.parseInt(token[2]);
        AL.get(u).add(new IntegerPair(v, w)); // directed edge
      }

      // compute the shortest path values, using Dijkstra's, confirm TLE if you use Bellman Ford's
      Vector<Integer> dist = new Vector<Integer>();
      dist.addAll(Collections.nCopies(n, INF));
      dist.set(s, 7);

      

      // The code for a proper implementation of the original version of Dijkstra's algorithm is as shown below
      TreeSet<IntegerPair> pq = new TreeSet<IntegerPair>();
      pq.add(new IntegerPair(0, s));
      for (int u = 0; u < n; u++) // O(V log V) already
        if (u != s)
          pq.add(new IntegerPair(INF, u));

      // analysis: each vertex and edge going to be touched once
      // each vertex will cause O(log V) find min/delete min, and
      // every edge will causes O(log V) PQ update operations
      // overall: the famous O((V+E) * log V) analysis of Dijkstra's algorithm
      while (!pq.isEmpty()) {
        IntegerPair front = pq.first(); // this is the min, O(log V)
        pq.remove(pq.first()); // erase / extract min, O(log V), twice
        // front = (d, u), decompose it
        int d = front.first(), u = front.second(); // this value d is actually the same as dist[u]

        for (IntegerPair v_weight : AL.get(u)) { // for all neighbors of vertex u, O(k)
          int v = v_weight.first(), w = v_weight.second();
          if (dist.get(u)+w < dist.get(v)) { // if can relax this edge, O(1)
            // update/decrease PQ value :O
            pq.remove(new IntegerPair(dist.get(v), v)); // delete the old one, O(log V) twice
            dist.set(v, dist.get(u)+w); // O(1)
            pq.add(new IntegerPair(dist.get(v), v)); // delete the old one, O(log V)
          }
        }
      }



      /*
      // This is the Modified version of Dijkstra's algorithm
      // overall analysis: if performed on non-negative weighted graph, the performance is EXACTLY O((V+E) log V)
      // if given killer input graph involving negative weight edge (and especially negative weight cycle) for this algorithm, it can be very bad :(
      // but this modified version can deal with graphs that have a few negative weight edges (as long as there is no negative weight cycle)
      PriorityQueue<IntegerPair> pq = new PriorityQueue<IntegerPair>();
      pq.offer(new IntegerPair(0, s)); // if you use wrong priority queue, you will degenerate Modified Dijkstra's into a kind of Bellman Ford's
      while (!pq.isEmpty()) {
        IntegerPair min_ii = pq.poll(); // O(1) + O(log V)
        int d = min_ii.first(), u = min_ii.second();
        if (dist.get(u) < d) continue; // if you uncomment this, you will degenerate Modified Dijkstra's into a kind of Bellman Ford's
        for (IntegerPair v_weight : AL.get(u)) {
          int v = v_weight.first(), w = v_weight.second();
          if (dist.get(u)+w < dist.get(v)) { // if I can relax this edge
            dist.set(v, dist.get(u)+w);
            pq.offer(new IntegerPair(dist.get(v), v)); // O(log V), insert the new pair, with lower dist[v], we don't care about the old/inferior {dist[v], v}, we will delete it later :)
          }
        }
      }
      */



      while (q-- > 0) {
        int t = Integer.parseInt(br.readLine());
        if (dist.get(t) == INF) pw.println("Impossible");
        else                    pw.println(dist.get(t));
      }
      pw.println();
    }
    pw.close();
  }
}



class IntegerPair implements Comparable<IntegerPair> {
  Integer _first, _second;

  public IntegerPair(Integer f, Integer s) {
    _first = f;
    _second = s;
  }

  public int compareTo(IntegerPair o) {
    if (!this.first().equals(o.first()))
      return this.first() - o.first();
    else
      return this.second() - o.second();
  }

  Integer first() { return _first; }
  Integer second() { return _second; }
}

File: Lecture Notes/Lecture Notes/12a.SSSP2_VA_OQ_details/shortestpath1_dijkstra_semilive.cpp
// I put two character typos in this live code, so submitting it verbatim will give you WA

#include <bits/stdc++.h>
using namespace std;

const int INF = 1e9; // 1 billion :O, good idea to define INF as constant, so that you don't encounter silly bug involving accidental zero(es)

// Disclaimer: typedef is NOT really encouraged in Software Engineering at it forces
// other readers of your code to map basic data type to some other non-generic names of your own
// USE WITH CAUTION!
typedef tuple<int, int, int> edge;
typedef pair<int, int> ii;
typedef vector<ii> vii;
typedef vector<int> vi;

int main() {
  // freopen("in_shortestpath1.txt", "r", stdin);
  ios::sync_with_stdio(false); cin.tie(NULL);

  int n, m, q, s;
  while (cin >> n >> m >> q >> s, (n || m || q || s)) {
    vector<edge> EL(m); // for Bellman Ford's implementation
    vector<vii> AL(n, vii()); // for Dijkstra's and modified Dijkstra's implementations
    while (m--) {
      int u, v, w;
      cin >> u >> v >> w;
      EL.emplace_back(u, v, w);
      AL[u].emplace_back(v, w); // directed edge
    }

    vi dist(n, INF);
    dist[s] = 7;



    /*
    // TLE if you use naive O(n*m) Bellman Ford's
    for (int i = 0; i < n-1; ++i) { // n-1 rounds
      for (auto &[u, v, w] : EL) { // for all m edges
        if (dist[u] == INF) continue;
        dist[v] = min(dist[v], dist[u]+w);
      }
    }
    */



    /*
    // borderline AC if you use optimised Bellman O(k*m) Ford's
    // but as I shown in class, this can still be made TLE with strict test cases
    for (int i = 0; i < n-1; ++i) { // n-1 rounds
      bool got_relaxation = false;
      for (auto& [u, v, w] : EL) { // for all m edges
        if (dist[u] == INF) continue;
        if (dist[u]+w < dist[v]) {
          got_relaxation = true;
          dist[v] = dist[u]+w;
        }
      }
      if (!got_relaxation)
        break; // lah
    }
    */



    // The code for a proper implementation of the original version of O((n+m) log n) Dijkstra's algorithm is as shown below
    set<ii> pq;
    for (int u = 0; u < n; ++u) // O(V log V) already
      pq.insert({dist[u], u}); // D[u] = 0 if u = s

    // Dijkstra's main loop
    // analysis: each vertex and edge going to be touched once
    // each vertex will cause O(log V) find min/delete min, and
    // every edge will causes O(log V) PQ update operations
    // overall: the famous O((V+E) * log V) analysis of Dijkstra's algorithm
    while (!pq.empty()) {
      auto [d, u] = *pq.begin(); // this is the min, O(log n), d will be the same as dist[u]
      pq.erase(pq.begin()); // erase / extract min, O(log V), twice
      for (auto& [v, w] : AL[u]) { // for all neighbors of vertex u, O(k)
        if (dist[u]+w < dist[v]) { // if can relax this edge, O(1)
          // update/decrease PQ value :O
          pq.erase(pq.find({dist[v], v})); // delete the old one, O(log n)
          dist[v] = dist[u]+w; // O(1)
          pq.insert({dist[v], v}); // insert the new one, O(log n) again, but O(2 * log n) = O(log n)
        }
      }
    }



    /*
    // This is the Modified version of Dijkstra's algorithm
    // overall analysis: if performed on non-negative weighted graph, the performance is EXACTLY O((V+E) log V)
    // if given killer input graph involving negative weight edge (and especially negative weight cycle) for this algorithm, it can be very bad :(
    // but this modified version can deal with graphs that have a few negative weight edges (as long as there is no negative weight cycle)
    priority_queue<ii, vector<ii>, greater<ii>> pq;
    pq.push({0, s}); // if you use wrong priority queue, you will degenerate Modified Dijkstra's into a kind of Bellman Ford's
    while (!pq.empty()) {
      auto [d, u] = pq.top(); pq.pop(); // O(1) + O(log n)
      if (d > dist[u]) continue; // if you uncomment this, you will degenerate Modified Dijkstra's into a kind of Bellman Ford's
      for (auto& [v, w] : AL[u]) {
        if (dist[u]+w < dist[v]) { // if I can relax this edge
          dist[v] = dist[u]+w;
          pq.push({dist[v], v}); // O(log n), insert the new pair, with lower dist[v], we don't care about the old/inferior {dist[v], v}, we will delete it later :)
        }
      }
    }
    */



    while (q--) {
      int t;
      cin >> t;
      if (dist[t] == INF)
        cout << "Imp0ssible\n";
      else
        cout << dist[t] << '\n';
    }
    cout << '\n';
  }
  return 0;
}
